Basic_viewer/include/CGAL/Graphics_scene.h

```diff
@@ -0,0 +1,420 @@
+    m_buffer_for_mono_rays.clear();
+    m_buffer_for_colored_rays.clear();
+    m_buffer_for_mono_lines.clear();
+    m_buffer_for_colored_lines.clear();
+    m_buffer_for_mono_faces.clear();
+    m_buffer_for_colored_faces.clear();
+    m_buffer_for_clipping_plane.clear();
+    m_texts.clear();
+    m_bounding_box=CGAL::Bbox_3();
+  }
+
+  template <typename KPoint>
+  static Local_point get_local_point(const KPoint &p)
+  {
+    return internal::Geom_utils<typename CGAL::Kernel_traits<KPoint>::Kernel,
+                                Local_kernel>::get_local_point(p);
+  }
+
+  template <typename KVector>
+  static Local_vector get_local_vector(const KVector &v)
+  {
+    return internal::Geom_utils<typename CGAL::Kernel_traits<KVector>::Kernel,
+                                Local_kernel>::get_local_vector(v);
+  }
+
+  template <typename KPoint>
+  void add_text(const KPoint &kp, const QString &txt)
+  {
+    Local_point p = get_local_point(kp);
+    m_texts.push_back(std::make_tuple(p, txt));
+  }
+
+  template <typename KPoint> void add_text(const KPoint &kp, const char *txt)
+  { add_text(kp, QString(txt)); }
+
+  template <typename KPoint>
+  void add_text(const KPoint &kp, const std::string &txt)
+  { add_text(kp, txt.c_str()); }
+
+  void m_texts_clear()
+  { m_texts.clear(); }
+
+  int m_texts_size()
+  { return m_texts.size(); }
+
+  std::vector<std::tuple<Local_point, QString>>& get_m_texts()
+  { return m_texts; }
+
+public:
+  // The following enum gives the indices of different elements of arrays
+  // vectors.
+  enum Buffers {
+    BEGIN_POS = 0,
+    POS_MONO_POINTS = BEGIN_POS,
+    POS_COLORED_POINTS,
+    POS_MONO_SEGMENTS,
+    POS_COLORED_SEGMENTS,
+    POS_MONO_RAYS,
+    POS_COLORED_RAYS,
+    POS_MONO_LINES,
+    POS_COLORED_LINES,
+    POS_MONO_FACES,
+    POS_COLORED_FACES,
+    POS_CLIPPING_PLANE,
+    END_POS,
+    BEGIN_COLOR = END_POS,
+    COLOR_POINTS = BEGIN_COLOR,
+    COLOR_SEGMENTS,
+    COLOR_RAYS,
+    COLOR_LINES,
+    COLOR_FACES,
+    END_COLOR,
+    BEGIN_NORMAL = END_COLOR,
+    SMOOTH_NORMAL_MONO_FACES = BEGIN_NORMAL,
+    FLAT_NORMAL_MONO_FACES,
+    SMOOTH_NORMAL_COLORED_FACES,
+    FLAT_NORMAL_COLORED_FACES,
+    END_NORMAL,
+    LAST_INDEX = END_NORMAL
+  };
+
+protected:
+  Buffer_for_vao<BufferType> m_buffer_for_mono_points;
+  Buffer_for_vao<BufferType> m_buffer_for_colored_points;
+  Buffer_for_vao<BufferType> m_buffer_for_mono_segments;
+  Buffer_for_vao<BufferType> m_buffer_for_colored_segments;
+  Buffer_for_vao<BufferType> m_buffer_for_mono_rays;
+  Buffer_for_vao<BufferType> m_buffer_for_colored_rays;
+  Buffer_for_vao<BufferType> m_buffer_for_mono_lines;
+  Buffer_for_vao<BufferType> m_buffer_for_colored_lines;
+  Buffer_for_vao<BufferType> m_buffer_for_mono_faces;
+  Buffer_for_vao<BufferType> m_buffer_for_colored_faces;
+  Buffer_for_vao<BufferType> m_buffer_for_clipping_plane;
+
+  std::vector<std::tuple<Local_point, QString>> m_texts;
+
+  std::vector<float> arrays[LAST_INDEX];
```

Comment:
`float`?! What if `BufferType`/`DataType` is not `float`?!

Reply:
this is a bug.